orangecontrib/prototypes/explanation/explainer.py

```diff
@@ -0,0 +1,568 @@
+        SHAP values sorted from most to least important, second contains
+        negative SHAP values sorted from most to least important.
+    segments
+        Value on y-axis to start/end segment. It is a list with a tuple for
+        each data item. Each tuple contains two arrays First are segments for
+        positive SHAP values sorted from most to least important, second
+        contains segments for negative SHAP values sorted from most to least
+        important. Each segment is a tuple with starting and the ending point.
+    selected_labels
+        Corresponding attributes and values. It is a list with a tuple for each
+        data item (row in data). Each tuple contains two arrays. In first are
+        attributes and values with positive SHAP values sorted from most to
+        least important, second contains attributes with negative SHAP values
+        sorted from most to least important.
+    ranges
+        Range of the graph for each data item.
+    """
+    shap_values = shap_values[target_class]
+    if top_n_features is None:
+        top_n_features = shap_values.shape[1]
+
+    top_features_idx = np.fliplr(np.argsort(np.abs(shap_values), axis=1),)[
+        :, :top_n_features
+    ]
+
+    assert top_features_idx.shape == (
+        len(shap_values),
+        min(shap_values.shape[1], top_n_features),
+    )
+
+    data_attributes = transformed_data.domain.attributes
+
+    selected_shap_values = []
+    selected_labels = []
+    segments = []
+    ranges = []
+    for sv, tfi, dat, pred in zip(
+        shap_values, top_features_idx, transformed_data, predictions
+    ):
+        positive_idx = tfi[sv[tfi] > 0]
+        negative_ids = tfi[sv[tfi] < 0]
+        selected_shap_values.append(
+            (sv[positive_idx].tolist(), sv[negative_ids].tolist())
+        )
+        selected_labels.append(
+            (
+                [(data_attributes[i].name, dat.x[i]) for i in positive_idx],
+                [(data_attributes[i].name, dat.x[i]) for i in negative_ids],
+            )
+        )
+        pos_segments = _compute_segments(sv[positive_idx], pred[target_class])
+        neg_segments = _compute_segments(sv[negative_ids], pred[target_class])
+        segments.append((pos_segments, neg_segments))
+        ranges.append(
+            (
+                pos_segments[-1][1]
+                if len(pos_segments)
+                else pred[target_class],
+                neg_segments[-1][1]
+                if len(neg_segments)
+                else pred[target_class],
+            )
+        )
+    return selected_shap_values, segments, selected_labels, ranges
+
+
+if __name__ == "__main__":
+    from Orange.classification import RandomForestLearner
+
+    data_ = Table.from_file("heart_disease.tab")
+    learner = RandomForestLearner()
+    model_ = learner(data_)
+
+    shap_val, transformed_domain, mask, colors_ = get_shap_values_and_colors(
+        model_, data_
+    )
```

Comment:
If I run the following code, it crashes:
```
    from Orange.classification import RandomForestLearner, LogisticRegressionLearner

    data_ = Table.from_file("heart_disease.tab")[:1]
    learner = LogisticRegressionLearner()
    model_ = learner(data_)

    shap_val, transformed_domain, mask, colors_ = get_shap_values_and_colors(
        model_, data_
    )
```

Reply:
It is fixed now.